README.md
# Rqmts

[![PyPI Download/Month](https://img.shields.io/pypi/dm/rqmts.svg)](https://pypi.python.org/pypi/rqmts/)
[![PyPI Version](https://img.shields.io/pypi/v/rqmts.svg)](https://pypi.python.org/pypi/rqmts/)
[![PyPI license](https://img.shields.io/pypi/l/rqmts.svg)](https://pypi.python.org/pypi/rqmts/)
[![PyPI pyversions](https://img.shields.io/pypi/pyversions/rqmts.svg)](https://pypi.python.org/pypi/rqmts/)
[![PyPI implementation](https://img.shields.io/pypi/implementation/rqmts.svg)](https://pypi.python.org/pypi/rqmts/)

<p align="center">
<b>Rqmts - Generate requirements.txt file for any project by analysing package imports</b><br><br>
    <img alt="Rqmts logo" src="https://i.imgur.com/czbQOUj.png" width="400"><br>
<b>Click <a href="#Usage">here</a> to see the demo.<br>
  Click <a href="https://github.com/0x48piraj/rqmts/wiki">here</a> for documentation.<br><br></b>
</p>

## About the project

**Rqmts** is a fantastic stand-alone tool which generates `requirements.txt` file for any project by analysing package imports.

It does not requires any dependency (works out-of-the-box), not needs internet to work _(is completely offline, upto this moment)_, nor uses regular expressions in such a violent way as existing projects do. Instead, it uses simple heuristic techniques and parse conditional trees, which is a better method for extracting imported names from statements, functions, etc.

## Why this project

### Questions

- **Why not just use pip's freeze command** to generate a `requirements.txt` file for my project?
- Why to **re-invent the wheel** when there are modules such as **pipreqs**, **pigar**, **poetry** already present?
- Why not manually?

### Answers

* **Why not just pip freeze?**
   * ``pip freeze`` only saves the packages that are installed with ``pip install`` in your environment. 
   * ``pip freeze`` saves all packages in the environment including those that you don't use in your current project. _(if you don't have virtualenv)_
* Why **re-invent the wheel** (?)
   * **pipreqs** fails on many occasions _(see - [pipreqs/issues](https://github.com/bndr/pipreqs/issues))_
   * **pigar** queries pypi servers, big no-no. Ideally, it should be local, on fallback? then maybe. Other than that, **pigar** recommends using Pipenv ([pipenv has serious issues](https://news.ycombinator.com/item?id=18612590))
   * **poetry** quotes "Be aware, however, that it will also install poetry's dependencies which might cause conflicts."
   * Sheer curiousity. _"can I create a project that has potential of collecting thosands of stars and most importantly, hundreds of contributors?"_
* Manually?
   * _Are you serious right now?_

## Installation


Use **pip** to install [rqmts](https://pypi.org/project/rqmts/). This is the **recommended way of running Rqmts**.

![Installation Guide](https://github.com/0x48piraj/rqmts/raw/master/static/installation-guide.gif)

**Rqmts** also provides a script that will run a **rqmts instance** isolated from the rest of your system by using file-less/memory-based execution.

```
curl -sSL https://raw.githubusercontent.com/0x48piraj/rqmts/master/Rqmts.py | python
```

Alternatively, you can download `Rqmts.py` from the root directory and execute it separately.

## Usage

#### Command-line Interface and Interactive mode

![Usage Demo](https://github.com/0x48piraj/rqmts/raw/master/static/usage-demo.gif)

#### Challenges

The major challenge of this project is to extract the required metadata from modules which are first extracted from the input script.

- **Version numbers in python can be in very different places depending on the case**
- **Package name in the package index is independent of the module name we import**

and these quirks make this project interesting. There's a funny comment in the source which reflects the diversity between us and it goes like :

```py
# module_name.__version__ sucks, because we suck (PEP 0396)
```

This project aims to combine the best existing strategies to cover the broadest possible set of cases _(if not all)_. The project was built keeping in mind the modular programming paradigms and so other than being readable it's easily extensible making it possible to add new strategies/algorithms quickly.

## Changelog

Playing with different versions? Read the **[Changelog](CHANGELOG.md)**.

## Contribute

All patches are Welcome! Please see [CONTRIBUTING.md](CONTRIBUTING.md) for further details. For discussions, create a [new issue](https://github.com/0x48piraj/rqmts/issues/new) or ping me over [Twitter](https://twitter.com/0x48piraj)

Again, if you have any issues or suggestions/patches, please do not hesitate to **[open an issue](https://github.com/0x48piraj/rqmts/issues/new)** or a **[pull request](https://github.com/0x48piraj/rqmts/pulls)**!

## Running the tests

This project uses [unittest](https://docs.python.org/2/library/unittest.html).

> The Python unit testing framework, sometimes referred to as "PyUnit," is a Python language version of JUnit, by Kent Beck and Erich Gamma. JUnit is, in turn, a Java version of Kent’s Smalltalk testing framework. Each is the de facto standard unit testing framework for its respective language.

You may need to install the package (rqmts) for setup beforehand, using

```
pip install --user rqmts
```

All tests are encapsulated in one single script named `testsuite.py` and all the respective test-cases are under `tests/testcases/`

For getting started,

```
python tests/testsuite.py
```

## License

This software is licensed under **BSD 3-Clause "New" or "Revised" License**. To view a copy of this license, visit **[BSD 3-Clause](LICENSE)**.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.0.3] - 2019-10-03

### Added

- Improved parsing sub-routine
- Improved project documentation
- New demonstration GIFs (Usage and Installation guide)
- New method for version detection via searching `__version__` file using AST (Abstract Syntax Trees). Works on Python > 3.4 as `importlib.util.find_spec` is not available in Python < 3.4
- `--verbose` flag for extended output

### Changed

- Moved `tests` to `testcases/`
- Updated `Rqmts.py` script
- Stdout process, fixed trailing newline via `.tell()` method and `os.linesep`

### Removed

- Old demos (Command-line & Interactive)

tests/test.py
import lxml
import re
import time
import string
import strgen
from random import randint
import os
from selenium.webdriver.common.desired_capabilities import DesiredCapabilities
from faker import Faker
from bs4 import BeautifulSoup
import pymouse
import chilkat
import sys
import requests, json
import idontexist

tests/testsuite.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys
import unittest
sys.path.append('..')
from rqmts.parsing import parse, fetch

class TestApp(unittest.TestCase):
   """Test the parse(), fetch() function"""

   def setup(self):
       """This runs before the test cases are executed"""

   def test_0001(self):
       """Test module parsing function"""
       script = open('testcases/test.py', 'r').read()
       result = parse(script)
       print("\n======================================================================\n")
       [print(num+1, module) for num, module in enumerate(result)]
       print("\n======================================================================\n")
       self.assertEqual(len(result), 16)
       
   def test_0002(self):
       """Test module package version fetching function"""
       script = open('testcases/test.py', 'r').read()
       result = parse(script)
       self.assertEqual(len(result), 16)
       requirements_list = []
       for i in result:
        try:
            pkg_name, version = fetch(i)
            res = pkg_name + '==' + version
            requirements_list.append(res)
        except Exception as e:
            pass
       self.assertEqual(len(requirements_list), 8)

def suite():
   """Test suite"""
   suite = unittest.TestSuite()
   suite.addTests(
       unittest.TestLoader().loadTestsFromTestCase(TestApp)
   )
   return suite

if __name__ == '__main__':
   unittest.TextTestRunner(verbosity=2).run(suite())

rqmts.py
#!/usr/bin/env python3

import os, sys, argparse, textwrap
from parsing import parse, fetch, _import

banner = textwrap.dedent('''\
    .===================================================================.
    ||      rqmts - Generate pip requirements.txt for any project      ||
    ||                                                                 ||
    ||                                                                 ||
    ||                            ___                                  ||
    ||                          .'   '.              requirements.txt  ||
    ||                         / rqmts \           oOoOo.              ||
    ||                        |    |    |       ,==|||||o.             ||
    ||                         \  pip  /       _|| |||||o. pip         ||
    ||                          '.___.'    _.-'^|| |||||               ||
    ||                        __/_______.-'     '==rqmts               ||
    ||                   _.-'` /                   """""               ||
    ||                .-'     /   oOoOo.                               ||
    ||                `-._   / ,==|||||o.                              ||
    ||                    '-/._|| |||||o. python                       ||
    ||                     /  ^|| |||||                                ||
    ||                    /    '==rqmts                                ||
    ||                   /________"""""                                ||
    ||                   `\       `\                                   ||
    ||                     \        `\                                 ||
    ||                      \         `\                               ||
    ||                      /           /                              ||
    ||                     /           /                               ||
    ||        @0x48piraj  /_____      /_____                           ||
    ||                                                                 ||
    '==================================================================='
    ''')

parser = argparse.ArgumentParser(formatter_class=argparse.RawDescriptionHelpFormatter, description=banner)
optional = parser._action_groups.pop() # popped opt args
optional = parser.add_argument_group('Options')
optional.add_argument("-p", "--path", dest="path", metavar="FILE", default=False, help= "Path of the Python script (inside quotation marks)")

file_path = parser.parse_args().path
print(banner)
if file_path == False:
    print("[*] Path not provided, invoking interactive mode ...")
    print("[*] Enter the path of Python script")
    file_path = input("    ----> ")

if os.path.exists(file_path):
 dir_path = os.path.dirname(file_path)
 with open(file_path, 'r') as f:
    try:
        code = f.read()
        f.close()
    except:
        print("[-] File failed to load. Exiting ...")
        sys.exit(1)
else:
 print("[-] Invalid path. Exiting ...")
 sys.exit(1)

requirements_list = []
REQS_PATH = dir_path + "\\requirements.txt"
PARSED_PKG_LIST = parse(code)
modules = _import(PARSED_PKG_LIST) # must run w/o errors or you've some fucked up imports
for package in PARSED_PKG_LIST:
    try:
        # Unfortunately, fetch() can error out too as name in the package index is independent of the module name we import
        pkg_name, version = fetch(package) # pkg.__version__ sucks, because we suck (PEP 0396)
        result = pkg_name + '==' + version
        requirements_list.append(result)
    except Exception as e:
        if e.__class__.__name__ == "ModuleNotFoundError":
            print("[!] ERROR: {} !(valid || installed)".format(package))
        elif e.__class__.__name__ == "DistributionNotFound":
            # hacky patch
            # pkg_name, version = fetch(dir(package)[0]) # works on pymouse, strgen
            print("[!] System package found :", package)

print("[+] Success: Parsed the dependencies correctly")
print("[*] Saving generated requirements.txt")

with open(REQS_PATH, 'w') as g:
    for req in requirements_list:
        g.write(req + '\n')
    g.close()

print("[+] Success: requirements.txt saved")
print("[+] Path where it can be found: %s" % (REQS_PATH))

rqmts/parsing.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import pkg_resources, dis, ast
import importlib.util
from collections import defaultdict

def _import(imports):
    modules = {}
    errors = []
    for module in imports:
        try:
            modules[module] = __import__(module)
        except ImportError:
            errors.append(module)
    return modules, errors

def find_version(PKG_DIR):
    """
    Return value of __version__ (works on Python > 3.4)
    Reference: https://stackoverflow.com/a/42269185/
    """
    file_path = importlib.util.find_spec(PKG_DIR).origin
    with open(file_path) as file_obj:
        root_node = ast.parse(file_obj.read())
    for node in ast.walk(root_node):
        if isinstance(node, ast.Assign):
            if len(node.targets) == 1 and node.targets[0].id == "__version__":
                return node.value.s
    raise RuntimeError("Unable to find __version__ string.")

def fetch(package):
    try:
        pkg_name, pkg_version = pkg_resources.get_distribution(package).project_name, pkg_resources.get_distribution(package).version
    except:
        pkg_name, pkg_version = package, find_version(package)
    return pkg_name, pkg_version

def parse(code):
    instructions = dis.get_instructions(code)
    imports = [__ for __ in instructions if 'IMPORT' in __.opname]
    grouped = defaultdict(list)
    for instr in imports:
        if '.' in instr.argval: # dirty code
            grouped[instr.opname].append(instr.argval.split('.')[0])
        else:
            grouped[instr.opname].append(instr.argval)
    return grouped['IMPORT_NAME']

rqmts/__main__.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import os, sys, argparse, textwrap
from .parsing import parse, fetch, _import

def main():
	if sys.platform.lower() == "win32":
	    os.system('color')
	  # Group of Different functions for different styles
	    class style():
	        BLACK = lambda x: '\033[30m' + str(x)
	        RED = lambda x: '\033[31m' + str(x)
	        GREEN = lambda x: '\033[32m' + str(x)
	        YELLOW = lambda x: '\033[33m' + str(x)
	        BLUE = lambda x: '\033[34m' + str(x)
	        MAGENTA = lambda x: '\033[35m' + str(x)
	        CYAN = lambda x: '\033[36m' + str(x)
	        WHITE = lambda x: '\033[37m' + str(x)
	        UNDERLINE = lambda x: '\033[4m' + str(x)
	        RESET = lambda x: '\033[0m' + str(x)
	else:
	    class style():
	        BLACK = ""
	        RED = ""
	        GREEN = ""
	        YELLOW = ""
	        BLUE = ""
	        MAGENTA = ""
	        CYAN = ""
	        WHITE = ""
	        UNDERLINE = ""
	        RESET = ""

	banner = textwrap.dedent('''\
    
    
    
    
    
RRRRRRRRRRRRRRRRR        QQQQQQQQQ     MMMMMMMM               MMMMMMMMTTTTTTTTTTTTTTTTTTTTTTT   SSSSSSSSSSSSSSS      
R::::::::::::::::R     QQ:::::::::QQ   M:::::::M             M:::::::MT:::::::::::::::::::::T SS:::::::::::::::S     
R::::::RRRRRR:::::R  QQ:::::::::::::QQ M::::::::M           M::::::::MT:::::::::::::::::::::TS:::::SSSSSS::::::S     
RR:::::R     R:::::RQ:::::::QQQ:::::::QM:::::::::M         M:::::::::MT:::::TT:::::::TT:::::TS:::::S     SSSSSSS     
  R::::R     R:::::RQ::::::O   Q::::::QM::::::::::M       M::::::::::MTTTTTT  T:::::T  TTTTTTS:::::S                 
  R::::R     R:::::RQ:::::O     Q:::::QM:::::::::::M     M:::::::::::M        T:::::T        S:::::S                 
  R::::RRRRRR:::::R Q:::::O     Q:::::QM:::::::M::::M   M::::M:::::::M        T:::::T         S::::SSSS              
  R:::::::::::::RR  Q:::::O     Q:::::QM::::::M M::::M M::::M M::::::M        T:::::T          SS::::::SSSSS         
  R::::RRRRRR:::::R Q:::::O     Q:::::QM::::::M  M::::M::::M  M::::::M        T:::::T            SSS::::::::SS       
  R::::R     R:::::RQ:::::O     Q:::::QM::::::M   M:::::::M   M::::::M        T:::::T               SSSSSS::::S      
  R::::R     R:::::RQ:::::O  QQQQ:::::QM::::::M    M:::::M    M::::::M        T:::::T                    S:::::S     
  R::::R     R:::::RQ::::::O Q::::::::QM::::::M     MMMMM     M::::::M        T:::::T                    S:::::S     
RR:::::R     R:::::RQ:::::::QQ::::::::QM::::::M               M::::::M      TT:::::::TT      SSSSSSS     S:::::S     
R::::::R     R:::::R QQ::::::::::::::Q M::::::M               M::::::M      T:::::::::T      S::::::SSSSSS:::::S     
R::::::R     R:::::R   QQ:::::::::::Q  M::::::M               M::::::M      T:::::::::T      S:::::::::::::::SS      
RRRRRRRR     RRRRRRR     QQQQQQQQ::::QQMMMMMMMM               MMMMMMMM      TTTTTTTTTTT       SSSSSSSSSSSSSSS        
                                 Q:::::Q                                                                             
                                  QQQQQQ                                                                             
                                                                                                                     
                                                                                                                     
                                                                                                                     
                                                                                                                     
                                   
	    ''')

	parser = argparse.ArgumentParser(formatter_class=argparse.RawDescriptionHelpFormatter, description=style.GREEN(banner) + style.RESET(''), usage=style.GREEN("rqmts ") + style.YELLOW("[-h] [-p '{}']").format(style.UNDERLINE("/path/to/file") + style.RESET('') + style.YELLOW('')) + style.RESET(''))
	optional = parser._action_groups.pop() # popped opt args
	optional = parser.add_argument_group('Options')
	optional.add_argument("-p", "--path", dest="path", metavar=style.CYAN("'/path/to/file'") + style.RESET(''), default=False, help= style.GREEN("Python script path (inside quotation marks)") + style.RESET(''))
	optional.add_argument("-v", "--verbose", dest="verbose", action='store_true', help= style.GREEN("Verbose output") + style.RESET(''))

	verbose = parser.parse_args().verbose
	file_path = parser.parse_args().path
	print(style.GREEN(banner) + style.RESET(''))
	try:
	    if file_path == False:
	        print(style.YELLOW("[*] Path not provided, invoking interactive mode ..."))
	        print("[*] Enter the path of Python script" + style.RESET(''))
	        file_path = input(style.GREEN("    ----> ") + style.RESET('')).strip()
	    else:
	        file_path = file_path.strip()
	except KeyboardInterrupt:
	    print('Interrupted.')
	    sys.exit(0) # http://tldp.org/LDP/abs/html/exitcodes.html#EXITCODESREF

	if os.path.exists(file_path):
	 dir_path = os.path.dirname(file_path)
	 with open(file_path, 'r') as f:
	    try:
	        code = f.read()
	        f.close()
	    except:
	        print(style.RED("[-] File failed to load. Exiting ...") + style.RESET(''))
	        sys.exit(1)
	else:
	 print(style.RED("[-] Invalid path. Exiting ...") + style.RESET(''))
	 sys.exit(1)

	requirements_list = []
	if dir_path == '': # bad code
	    REQS_PATH = "requirements.txt"
	else:
	    REQS_PATH = dir_path + "\\requirements.txt"
	PARSED_PKG_LIST = parse(code)
	modules, errors = _import(PARSED_PKG_LIST)
	if len(PARSED_PKG_LIST) != len(modules) or len(errors) != 0:
	    ERROR_MSG = (
                '\n\n'
                'Some modules were not successfully imported, and it means either of the two things :\n'
                '1. No such module(s) (just ignore this warning then)\n'
                '2. Module(s) not found in your system. (pip install <module-name>)'
            )
	    print(style.YELLOW(ERROR_MSG) + style.RESET(''))
	    print(style.YELLOW('There seems to be discrepancies in following module(s) :') + style.RESET(''))
	    for num, module in enumerate(errors):
	        print(num + 1, module)
	    print(style.RED('Quitting.') + style.RESET(''))
	    sys.exit(0)
	
	"""Yes, we're not using https://github.com/jackmaney/python-stdlib-list"""
	for name, data in zip(modules.keys(), modules.values()): # removing built-in modules from PARSED_PKG_LIST
	    if "(built-in)" in str(data): # dirty, but reliable
	        print(style.RED("[!] System (built-in) package found :"), style.CYAN(style.UNDERLINE(name)) + style.RESET(''))
	        try:
	            PARSED_PKG_LIST.remove(name)
	        except ValueError:
	            pass
	    elif "\\lib\\" in str(data):
	        print(style.RED("[!] Library package found :"), style.CYAN(style.UNDERLINE(name)) + style.RESET(''))
	        try:
	            PARSED_PKG_LIST.remove(name)
	        except ValueError:
	            pass
	print(style.YELLOW("[+] Total modules used in %s => %d" % (style.CYAN(style.UNDERLINE(file_path)) + style.RESET(''), len(PARSED_PKG_LIST))) + style.RESET(''))
	print(style.YELLOW("[*] Fetching versions ...") + style.RESET(''))
	for package in PARSED_PKG_LIST:
	    try:
	        # Unfortunately, fetch() can error out too as name in the package index is independent of the module name we import
	        pkg_name, version = fetch(package) # pkg.__version__ sucks, because we suck (PEP 0396)
	        result = pkg_name + '==' + version
	        requirements_list.append(result)
	    except Exception as e:
	        print(style.RED("Couldn't find version of {}".format(style.UNDERLINE(package) + style.RESET(''))) + style.RESET(''))
	        if verbose:
	            print(style.RED("[Verbose] Error: {}".format(e)) + style.RESET(''))
	if len(requirements_list) != len(PARSED_PKG_LIST):
	    print(style.RED("[!] Failed to extract all versions.") + style.RESET(''))
	print(style.GREEN("[+] Success: Parsed all the dependencies") + style.RESET(''))
	print(style.YELLOW("[*] Saving generated ") + style.UNDERLINE("requirements.txt") + style.RESET(''))
	try:
	    with open(REQS_PATH, 'w') as g:
	        for req in requirements_list:
	            g.write(req + '\n')
	        # tell() gives current byte number in the file, we subtract one, truncate it to remove the trailing newline
	        g.truncate(g.tell() - len(os.linesep))
	        # On Linux and MacOS, the -1 is correct, on Windows it needs to be -2. Pythonic method, using os.linesep
	        g.close()
	except PermissionError:
		print(style.RED('Quitting. No permission to write on {}'.format(REQS_PATH)) + style.RESET(''))
		sys.exit(0)
	print(style.GREEN("[+] Success: ") + style.GREEN(style.UNDERLINE("requirements.txt")) + style.RESET('') + style.GREEN(" saved") + style.RESET(''))
	print(style.GREEN("[+] Path where it can be found: %s") % (style.GREEN(style.UNDERLINE(REQS_PATH))) + style.RESET(''))

if __name__ == "__main__":
    main()


Rqmts.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import os, sys, argparse, textwrap
import pkg_resources, dis
import ast
import importlib.util
from collections import defaultdict

def _import(imports):
    modules = {}
    errors = []
    for module in imports:
        try:
            modules[module] = __import__(module)
        except ImportError:
            errors.append(module)
    return modules, errors

def find_version(PKG_DIR):
    """
    Return value of __version__ (works on Python > 3.4)
    Reference: https://stackoverflow.com/a/42269185/
    """
    file_path = importlib.util.find_spec(PKG_DIR).origin
    with open(file_path) as file_obj:
        root_node = ast.parse(file_obj.read())
    for node in ast.walk(root_node):
        if isinstance(node, ast.Assign):
            if len(node.targets) == 1 and node.targets[0].id == "__version__":
                return node.value.s
    raise RuntimeError("Unable to find __version__ string.")

def fetch(package):
    try:
        pkg_name, pkg_version = pkg_resources.get_distribution(package).project_name, pkg_resources.get_distribution(package).version
    except:
        pkg_name, pkg_version = package, find_version(package)
    return pkg_name, pkg_version

def parse(code):
    instructions = dis.get_instructions(code)
    imports = [__ for __ in instructions if 'IMPORT' in __.opname]
    grouped = defaultdict(list)
    for instr in imports:
        if '.' in instr.argval:
            grouped[instr.opname].append(instr.argval.split('.')[0])
        else:
            grouped[instr.opname].append(instr.argval)
    return grouped['IMPORT_NAME']

if sys.platform.lower() == "win32":
    os.system('color')
  # Group of Different functions for different styles
    class style():
        BLACK = lambda x: '\033[30m' + str(x)
        RED = lambda x: '\033[31m' + str(x)
        GREEN = lambda x: '\033[32m' + str(x)
        YELLOW = lambda x: '\033[33m' + str(x)
        BLUE = lambda x: '\033[34m' + str(x)
        MAGENTA = lambda x: '\033[35m' + str(x)
        CYAN = lambda x: '\033[36m' + str(x)
        WHITE = lambda x: '\033[37m' + str(x)
        UNDERLINE = lambda x: '\033[4m' + str(x)
        RESET = lambda x: '\033[0m' + str(x)
else:
    class style():
        BLACK = ""
        RED = ""
        GREEN = ""
        YELLOW = ""
        BLUE = ""
        MAGENTA = ""
        CYAN = ""
        WHITE = ""
        UNDERLINE = ""
        RESET = ""

banner = textwrap.dedent('''\
    .===================================================================.
    ||      rqmts - Generate pip requirements.txt for any project      ||
    ||                                                                 ||
    ||                                                                 ||
    ||                            ___                                  ||
    ||                          .'   '.              requirements.txt  ||
    ||                         / rqmts \           oOoOo.              ||
    ||                        |    |    |       ,==|||||o.             ||
    ||                         \  pip  /       _|| |||||o. pip         ||
    ||                          '.___.'    _.-'^|| |||||               ||
    ||                        __/_______.-'     '==rqmts               ||
    ||                   _.-'` /                   """""               ||
    ||                .-'     /   oOoOo.                               ||
    ||                `-._   / ,==|||||o.                              ||
    ||                    '-/._|| |||||o. python                       ||
    ||                     /  ^|| |||||                                ||
    ||                    /    '==rqmts                                ||
    ||                   /________"""""                                ||
    ||                   `\       `\                                   ||
    ||                     \        `\                                 ||
    ||                      \         `\                               ||
    ||                      /           /                              ||
    ||                     /           /                               ||
    ||     [>] @0x48piraj /_____      /_____                           ||
    ||                                                                 ||
    '==================================================================='
    ''')

parser = argparse.ArgumentParser(formatter_class=argparse.RawDescriptionHelpFormatter, description=style.GREEN(banner) + style.RESET(''), usage=style.GREEN("rqmts ") + style.YELLOW("[-h] [-p '{}']").format(style.UNDERLINE("/path/to/file") + style.RESET('') + style.YELLOW('')) + style.RESET(''))
optional = parser._action_groups.pop() # popped opt args
optional = parser.add_argument_group('Options')
optional.add_argument("-p", "--path", dest="path", metavar=style.CYAN("'/path/to/file'") + style.RESET(''), default=False, help= style.GREEN("Python script path (inside quotation marks)") + style.RESET(''))
optional.add_argument("-v", "--verbose", dest="verbose", action='store_true', help= style.GREEN("Verbose output") + style.RESET(''))

verbose = parser.parse_args().verbose
file_path = parser.parse_args().path
print(style.GREEN(banner) + style.RESET(''))
try:
    if file_path == False:
        print(style.YELLOW("[*] Path not provided, invoking interactive mode ..."))
        print("[*] Enter the path of Python script" + style.RESET(''))
        file_path = input(style.GREEN("    ----> ") + style.RESET('')).strip()
    else:
        file_path = file_path.strip()
except KeyboardInterrupt:
    print('Interrupted.')
    sys.exit(0) # http://tldp.org/LDP/abs/html/exitcodes.html#EXITCODESREF

if os.path.exists(file_path):
 dir_path = os.path.dirname(file_path)
 with open(file_path, 'r') as f:
    try:
        code = f.read()
        f.close()
    except:
        print(style.RED("[-] File failed to load. Exiting ...") + style.RESET(''))
        sys.exit(1)
else:
 print(style.RED("[-] Invalid path. Exiting ...") + style.RESET(''))
 sys.exit(1)

requirements_list = []
if dir_path == '': # bad code
    REQS_PATH = "requirements.txt"
else:
    REQS_PATH = dir_path + "\\requirements.txt"
PARSED_PKG_LIST = parse(code)
modules, errors = _import(PARSED_PKG_LIST)
if len(PARSED_PKG_LIST) != len(modules) or len(errors) != 0:
    ERROR_MSG = (
                '\n\n'
                'Some modules were not successfully imported, and it means either of the two things :\n'
                '1. No such module(s) (just ignore this warning then)\n'
                '2. Module(s) not found in your system. (pip install <module-name>)'
            )
    print(style.YELLOW(ERROR_MSG) + style.RESET(''))
    print(style.YELLOW('There seems to be discrepancies in following module(s) :') + style.RESET(''))
    for num, module in enumerate(errors):
        print(num + 1, module)
    print(style.RED('Quitting.') + style.RESET(''))
    sys.exit(0)

"""Yes, we're not using https://github.com/jackmaney/python-stdlib-list"""
for name, data in zip(modules.keys(), modules.values()): # removing built-in modules from PARSED_PKG_LIST
    if "(built-in)" in str(data): # dirty, but reliable
        print(style.RED("[!] System (built-in) package found :"), style.CYAN(style.UNDERLINE(name)) + style.RESET(''))
        try:
            PARSED_PKG_LIST.remove(name)
        except ValueError:
            pass
    elif "\\lib\\" in str(data):
        print(style.RED("[!] Library package found :"), style.CYAN(style.UNDERLINE(name)) + style.RESET(''))
        try:
            PARSED_PKG_LIST.remove(name)
        except ValueError:
            pass
print(style.YELLOW("[+] Total modules used in %s => %d" % (style.CYAN(style.UNDERLINE(file_path)) + style.RESET(''), len(PARSED_PKG_LIST))) + style.RESET(''))
print(style.YELLOW("[*] Fetching versions ...") + style.RESET(''))
for package in PARSED_PKG_LIST:
    try:
        # Unfortunately, fetch() can error out too as name in the package index is independent of the module name we import
        pkg_name, version = fetch(package) # pkg.__version__ sucks, because we suck (PEP 0396)
        result = pkg_name + '==' + version
        requirements_list.append(result)
    except Exception as e:
        print(style.RED("Couldn't find version of {}".format(style.UNDERLINE(package) + style.RESET(''))) + style.RESET(''))
        if verbose:
            print(style.RED("[Verbose] Error: {}".format(e)) + style.RESET(''))
if len(requirements_list) != len(PARSED_PKG_LIST):
    print(style.RED("[!] Failed to extract all versions.") + style.RESET(''))
print(style.GREEN("[+] Success: Parsed all the dependencies") + style.RESET(''))
print(style.YELLOW("[*] Saving generated ") + style.UNDERLINE("requirements.txt") + style.RESET(''))
try:
    with open(REQS_PATH, 'w') as g:
        for req in requirements_list:
            g.write(req + '\n')
        # removing the trailing newline, https://stackoverflow.com/a/18139508
        g.truncate(g.tell() - len(os.linesep))
        g.close()
except PermissionError:
    print(style.RED('Quitting. No permission to write on {}'.format(REQS_PATH)) + style.RESET(''))
    sys.exit(0)
print(style.GREEN("[+] Success: ") + style.GREEN(style.UNDERLINE("requirements.txt")) + style.RESET('') + style.GREEN(" saved") + style.RESET(''))
print(style.GREEN("[+] Path where it can be found: %s") % (style.GREEN(style.UNDERLINE(REQS_PATH))) + style.RESET(''))